### PR TITLE
Seperate segmented algorithms for count and count_if

### PR DESCRIPTION
**Pull request by Jedi18**
Seperate segmented algorithms for count and count_if
Replace tag_invoke with tag_fallback_invoke

## Any background context you want to provide?
Issue #5156
Issue #5204

### CODE REVIEW

**Comment by hkaiser** [comment]
retest

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/count.hpp [hunk 1]
```diff
@@ -528,21 +480,25 @@ namespace hpx {
             )>
         // clang-format on
         friend typename std::iterator_traits<FwdIter>::difference_type
-        tag_invoke(count_t, FwdIter first, FwdIter last, T const& value)
+        tag_fallback_invoke(
+            count_t, FwdIter first, FwdIter last, T const& value)
         {
             static_assert((hpx::traits::is_forward_iterator<FwdIter>::value),
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/count.hpp:486]: Could you please change this to `is_input_iterator` while you're looking at this algorithm?

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/count.hpp [hunk 2]
```diff
@@ -321,12 +330,12 @@ namespace hpx { namespace ranges {
                 (hpx::traits::is_forward_iterator<iterator_type>::value),
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/count.hpp:330]: `is_input_iterator here as well.

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/count.hpp [hunk 3]
```diff
@@ -336,24 +345,26 @@ namespace hpx { namespace ranges {
                 hpx::traits::is_sentinel_for<Sent, Iter>::value
             )>
         // clang-format on
-        friend typename std::iterator_traits<Iter>::difference_type tag_invoke(
-            count_t, Iter first, Sent last, T const& value,
+        friend typename std::iterator_traits<Iter>::difference_type
+        tag_fallback_invoke(count_t, Iter first, Sent last, T const& value,
             Proj&& proj = Proj())
         {
             static_assert((hpx::traits::is_forward_iterator<Iter>::value),
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/count.hpp:352]: and here as well: `is_input_iterator`

File: libs/parallelism/algorithms/tests/unit/algorithms/countif_tests.hpp [hunk 4]
```diff
@@ -77,7 +77,11 @@ void test_count_if(ExPolicy&& policy, IteratorTag)
     diff_type num_items = hpx::count_if(policy, iterator(std::begin(c)),
         iterator(std::end(c)), smaller_than_50());
 
+    diff_type num_items1 = hpx::parallel::count_if(policy,
```
Comment (hkaiser) [inline libs/parallelism/algorithms/tests/unit/algorithms/countif_tests.hpp:80]: This probably should not call the old deprecated algorithm. Same below.

Comment (Jedi18) [reply libs/parallelism/algorithms/tests/unit/algorithms/countif_tests.hpp:80]: Oh ok I'll just remove the tests then.

**Comment by hkaiser** [comment]
@Jedi18 the CircleCI builder is complaining about a couple of unused `using` directives here: https://app.circleci.com/pipelines/github/STEllAR-GROUP/hpx/6926/workflows/a28cbe9c-935e-434d-89c2-a7601ceafdc9/jobs/241207/steps

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Comment by msimberg** [comment]
retest

**Pull request merged**